PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterPlayerScripts/ControlScript/MasterControl.rbxmx

```diff
@@ -12,95 +12,115 @@
 	// Description: All character control scripts go thru this script, this script makes sure all actions are performed
 --]]
 
+-- [[ Constants ]]--
+local ZERO_VECTOR3 = Vector3.new(0, 0, 0)
+local STATE_JUMPING = Enum.HumanoidStateType.Jumping
+local STATE_FREEFALL = Enum.HumanoidStateType.Freefall
+local STATE_LANDED = Enum.HumanoidStateType.Landed
+
 --[[ Local Variables ]]--
 local MasterControl = {}
 
 local Players = game:GetService('Players')
 local RunService = game:GetService('RunService')
-local UserInputService = game:GetService('UserInputService')
-
-local HasVRAPI = pcall(function() return UserInputService.VREnabled and UserInputService.GetUserCFrame end)
+local VRService = game:GetService('VRService')
+local VREnabled = VRService.VREnabled
 
 while not Players.LocalPlayer do
-	wait()
+	Players.PlayerAdded:wait()
 end
 local LocalPlayer = Players.LocalPlayer
+local LocalCharacter = LocalPlayer.Character
 local CachedHumanoid = nil
-local RenderSteppedCon = nil
-local SeatedCn = nil
-local moveFunc = LocalPlayer.Move
 
 local isJumping = false
-local isSeated = false
-local myVehicleSeat = nil
-local moveValue = Vector3.new(0,0,0)
-
+local moveValue = Vector3.new(0, 0, 0)
 
 --[[ Local Functions ]]--
 function MasterControl:GetHumanoid()
-	local character = LocalPlayer and LocalPlayer.Character
-	if character then
-		if CachedHumanoid and CachedHumanoid.Parent == character then
+	if LocalCharacter then
+		if CachedHumanoid then
 			return CachedHumanoid
 		else
-			CachedHumanoid = nil
-			for _,child in pairs(character:GetChildren()) do
-				if child:IsA('Humanoid') then
-					CachedHumanoid = child
-					return CachedHumanoid
-				end
-			end
+			CachedHumanoid = LocalCharacter:FindFirstChildOfClass("Humanoid")
+			return CachedHumanoid
 		end
 	end
 end
 
+VRService.Changed:connect(function(prop)
+	if prop == "VREnabled" then
+		VREnabled = VRService.VREnabled
+	end
+end)
+
+local characterAncestryChangedConn = nil
+local characterChildRemovedConn = nil
+local function characterAdded(character)
+	if characterAncestryChangedConn then
+		characterAncestryChangedConn:disconnect()
+	end
+
+	if characterChildRemovedConn then
+		characterChildRemovedConn:disconnect()
+	end	
+	
+	LocalCharacter = character
+	CachedHumanoid = LocalCharacter:FindFirstChildOfClass("Humanoid")
+	characterAncestryChangedConn = character.AncestryChanged:connect(function()
+		if not character:IsDescendantOf(workspace) then
```

Comment:
when does this happen? do we need to worry about this case? should this really be, if the character has no parent, lets stop tracking it?